storage/src/checker.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+use crate::range_set::LinearAddressRangeSet;
+use crate::{
+    CheckerError, Committed, HashedNodeReader, LinearAddress, Node, NodeReader, NodeStore,
+    WritableStorage,
+};
+
+/// [`NodeStore`] checker
+// TODO: S needs to be writeable if we ask checker to fix the issues
+impl<S: WritableStorage> NodeStore<Committed, S> {
+    /// Go through the filebacked storage and check for any inconsistencies. It proceeds in the following steps:
+    /// 1. Check the header
+    /// 2. traverse the trie and check the nodes
+    /// 3. check the free list
+    /// 4. check missed areas - what are the spaces between trie nodes and free lists we have traversed?
+    /// # Errors
+    /// Returns a [`CheckerError`] if the database is inconsistent.
+    // TODO: report all errors, not just the first one
+    // TODO: add merkle hash checks as well
+    pub fn check(&self) -> Result<(), CheckerError> {
```

Comment:
High level: this only returns a single error and only the first error we find. Ideally, we should be returning every error we can discover and only abort the entire scan when we find an error we cannot skip past. And even then, we should include all the errors we discovered up until that point. If there is something we can fix, then we will want to do that in bulk instead of (1) running fsck, (2) getting and error, (3) fixing it, and (1) running fsck again until (4) we get an error we can't fix or it's clean.

This could potentially be done using a callback instead of collecting everything into a vec. Have the caller of `check` pass in something that implements a trait that receives each error we discover as we discover it.